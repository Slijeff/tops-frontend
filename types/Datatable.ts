export interface tableDataRaw {
    jsontimestamp: string,
    eventstate: string,
    minendtime: string,
    signalgroupid: string,
    intersectionid: number
}

export interface tableData {
    jsontimestamp: string,
    eventstate: string,
    minendtime: string,
    movement: string,
    intersectionid: number
}

export interface DatatableProps{
    data: tableData[];
}