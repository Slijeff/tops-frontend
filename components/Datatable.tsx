import React from "react";
import { Box, Badge, Text } from "@chakra-ui/react";
import { DatatableProps } from "../types/Datatable";
import {
  DataGrid,
  GridRowsProp,
  GridColDef,
  GridToolbar,
  GridCellParams,
} from "@mui/x-data-grid";
import type {} from "@mui/x-data-grid/themeAugmentation";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import NoRowsDisaply from "./NoRowsDisplay";

const Datatable = ({ data, mode }: DatatableProps) => {
  const theme = createTheme({});

  const rows: GridRowsProp = data.map((d, i) => {
    return {
      id: i,
      Timestamp: d.jsontimestamp,
      Movement: d.movement,
      Status: d.eventstate,
      Endtime: d.minendtime,
    };
  });

  const columns: GridColDef[] = [
    { field: "Timestamp", flex: 1 },
    { field: "Movement", flex: 1 },
    { field: "Status", flex: 1 },
    { field: "Endtime", headerName: "End time", flex: 1 },
  ];

  return (
    <>
      {mode === "real" && (
        <>
          <Text fontSize={"xl"} fontWeight="bold">
            Record Timestamp:
          </Text>
          <Badge fontSize={"2em"} colorScheme="gray" rounded={"lg"}>
            {data[0] != undefined && data[0].jsontimestamp}
          </Badge>
        </>
      )}

      <ThemeProvider theme={theme}>
        <Box height={"600px"} width={"100%"} rounded={"2xl"}>
          <DataGrid
            rows={rows}
            columns={columns}
            components={{ Toolbar: GridToolbar, NoRowsOverlay: NoRowsDisaply }}
            sx={{
              "& .MuiDataGrid-columnHeaderTitle": { fontWeight: "bold" },
              borderRadius: "10px",
              "& .stop": { color: "#F56565" },
              "& .perm": { color: "#48BB78" },
              "& .prot": { color: "#2F855A" },
              "& .unkown": { color: "#A0AEC0" },
            }}
            disableColumnMenu
            getCellClassName={(params: GridCellParams<string>) => {
              if (params.value === "stop-And-Remain") {
                return "stop";
              } else if (params.value === "permissive-Movement-Allowed") {
                return "perm";
              } else if (params.value === "protected-Movement-Allowed") {
                return "prot";
              } else if (params.field === "Status") {
                return "unkown";
              }
              return "";
            }}
            columnVisibilityModel={{
              Timestamp: mode !== "real",
            }}
          />
        </Box>
      </ThemeProvider>
    </>
  );
};

export default Datatable;
