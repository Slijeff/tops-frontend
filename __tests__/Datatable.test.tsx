import {describe, expect} from "@jest/globals";
import Datatable from "../components/Datatable";
import {render} from "@testing-library/react";

describe("<Datatable/> rendering", () => {
    test("should render Datatable correctly", async () => {
        const {getByText, container, getAllByText} = render(<Datatable/>);
        expect(container.firstChild).toMatchSnapshot();
        expect(getByText("2019010321")).toBeTruthy();
        expect(getAllByText(/stop/i)).toBeTruthy();
        expect(getByText(/pass/i)).toBeTruthy();
        expect(getByText("1")).toBeTruthy();
    })
})