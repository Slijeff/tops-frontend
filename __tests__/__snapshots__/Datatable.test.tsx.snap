// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Datatable/> rendering should render Datatable correctly 1`] = `
<div
  class="chakra-table__container css-d81vpl"
>
  <table
    class="chakra-table css-0"
    role="table"
  >
    <thead
      class="css-0"
    >
      <tr
        class="css-0"
        role="row"
      >
        <th
          class="css-0"
        >
          Timestamp
        </th>
        <th
          class="css-0"
        >
          Signal Group
        </th>
        <th
          class="css-0"
        >
          Status
        </th>
        <th
          class="css-0"
        >
          End time
        </th>
      </tr>
    </thead>
    <tbody
      class="css-0"
    >
      <tr
        class="css-0"
        role="row"
      >
        <td
          class="css-0"
          role="gridcell"
        >
          2019010321
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          1
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          stop
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          32492840295
        </td>
      </tr>
      <tr
        class="css-0"
        role="row"
      >
        <td
          class="css-0"
          role="gridcell"
        >
          2019010322
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          2
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          Pass
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          32492840295
        </td>
      </tr>
      <tr
        class="css-0"
        role="row"
      >
        <td
          class="css-0"
          role="gridcell"
        >
          2019010323
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          3
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          stop
        </td>
        <td
          class="css-0"
          role="gridcell"
        >
          32492840295
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
